lib/honeybadger/insights/base.ex

```diff
@@ -0,0 +1,164 @@
+defmodule Honeybadger.Insights.Base do
+  @moduledoc """
+  Base module providing common telemetry attachment functionality.
+  """
+
+  defmacro __using__(_opts) do
+    quote do
+      require Logger
+
+      # Define empty defaults for the attributes
+      Module.register_attribute(__MODULE__, :required_dependencies, [])
+      @required_dependencies []
+      Module.register_attribute(__MODULE__, :telemetry_events, [])
+      @time_keys ~w(duration total_time decode_time query_time queue_time idle_time)a
+
+      @before_compile Honeybadger.Insights.Base
+
+      def dependencies_available? do
+        Honeybadger.Insights.Base.dependencies_available?(@required_dependencies)
+      end
+
+      def get_insights_config(key, default) do
+        Honeybadger.Insights.Base.get_insights_config(key, default, __MODULE__)
+      end
+    end
+  end
+
+  @doc """
+  Checks if all required dependencies are available.
+  """
+  def dependencies_available?(deps) do
+    if Enum.empty?(deps), do: true, else: Enum.all?(deps, &Code.ensure_loaded?/1)
+  end
+
+  @doc """
+  Retrieves a configuration value from the insights configuration.
+  """
+  def get_insights_config(key, default, mod) do
+    config_namespace =
+      mod
+      |> Atom.to_string()
+      |> String.split(".")
+      |> List.last()
+      |> Macro.underscore()
+      |> String.to_atom()
+
+    insights_config = Application.get_env(:honeybadger, :insights_config, %{})
+    module_config = Map.get(insights_config, config_namespace, %{})
+    Map.get(module_config, key, default)
+  end
+
+  defmacro __before_compile__(_env) do
+    quote do
+      def event_filter(map, meta, name) do
+        if Application.get_env(:honeybadger, :event_filter) do
+          Application.get_env(:honeybadger, :event_filter).filter(map, meta, name)
+        else
+          map
+        end
+      end
+
+      def get_telemetry_events() do
+        get_insights_config(:telemetry_events, @telemetry_events)
+      end
+
+      @doc """
+      Attaches telemetry handlers if all required dependencies are available.
+      """
+      def attach do
+        if dependencies_available?() do
+          get_telemetry_events()
+          |> Enum.each(&attach_event/1)
```

Comment:
Same "single pipe" nitpick:

```suggestion
          Enum.each(get_telemetry_events(), &attach_event/1)
```